### PR TITLE
[GEOT-7019] Appschema support for geopackage

### PR DESCRIPTION
**Pull request by ydeliorman**
[GEOT-7019](https://osgeo-org.atlassian.net/browse/GEOT-7019) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

Appschema integration is done for Geopackage
  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [x] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [x] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [x] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [ ] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

File: modules/plugin/geopkg/src/main/java/org/geotools/geopkg/GeoPkgDataStoreFactory.java [hunk 1]
```diff
@@ -106,6 +109,20 @@ protected String getDriverClassName() {
         return "org.sqlite.JDBC";
     }
 
+    @Override
+    protected SQLDialect createSQLDialect(JDBCDataStore dataStore, Map<String, ?> params) {
```
Comment (aaime) [inline modules/plugin/geopkg/src/main/java/org/geotools/geopkg/GeoPkgDataStoreFactory.java:113]: Please add a test in this module (and then one to show the app-schema integration is working)

Comment (ydeliorman) [reply modules/plugin/geopkg/src/main/java/org/geotools/geopkg/GeoPkgDataStoreFactory.java:113]: The agreement with @nmco and @fernandor777 to put the implementation first then to put integration and unit tests.

Comment (nmco) [reply modules/plugin/geopkg/src/main/java/org/geotools/geopkg/GeoPkgDataStoreFactory.java:113]: @ydeliorman no PR is merged without and test.

File: modules/plugin/geopkg/src/test/java/org/geotools/geopkg/GeoPkgDataStoreFactoryTest.java [hunk 2]
```diff
@@ -57,4 +59,30 @@ public void testBaseDirectory() throws IOException {
...
+        b.setName("foo");
+        b.setNamespaceURI("http://geotools.org");
+        b.setSRS("EPSG:4326");
+        b.add("geom", Point.class);
+        b.add("name", String.class);
+
+        JDBCDataStore data = factory.createDataStore(map);
+        data.createSchema(b.buildFeatureType());
+
+        assertTrue(
```
Comment (aaime) [inline modules/plugin/geopkg/src/test/java/org/geotools/geopkg/GeoPkgDataStoreFactoryTest.java:84]: Wouldn't this test pass even without providing the "contents_only" parameter?
A valid test must fail before the fix.
I was expecting to see the direct creation of a table, using direct SQL, that is, a table that would not be registered in gpkg_contents (createSchema should do that instead).

Comment (ydeliorman) [reply modules/plugin/geopkg/src/test/java/org/geotools/geopkg/GeoPkgDataStoreFactoryTest.java]: done

File: modules/plugin/geopkg/src/test/java/org/geotools/geopkg/GeoPkgDataStoreContentsOnlyTest.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ *    Lesser General Public License for more details.
+ */
+package org.geotools.geopkg;
+
+import java.sql.SQLException;
+import org.geotools.jdbc.JDBCBooleanOnlineTest;
+import org.geotools.jdbc.JDBCBooleanTestSetup;
+import org.junit.Test;
+
+public class GeoPkgDataStoreContentsOnlyTest extends JDBCBooleanOnlineTest {
```
Comment (aaime) [inline modules/plugin/geopkg/src/test/java/org/geotools/geopkg/GeoPkgDataStoreContentsOnlyTest.java:24]: Good, but should not extend another test class with tests: you inherit those tests as well and they get run twice.
Instead:

```suggestion
public class GeoPkgDataStoreContentsOnlyTest extends JDBCTestSupport {
```

File: modules/plugin/geopkg/src/test/java/org/geotools/geopkg/GeoPkgDataStoreFactoryTestSetup.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ */
+package org.geotools.geopkg;
+
+import org.geotools.jdbc.JDBCBooleanTestSetup;
+import org.geotools.jdbc.JDBCTestSetup;
+
+public class GeoPkgDataStoreFactoryTestSetup extends JDBCBooleanTestSetup {
```
Comment (aaime) [inline modules/plugin/geopkg/src/test/java/org/geotools/geopkg/GeoPkgDataStoreFactoryTestSetup.java:22]:
```suggestion
public class GeoPkgDataStoreFactoryTestSetup extends JDBCDelegatingTestSetup {
```

You'll also have to adjust the methods below.

These two classes show an example of a test that is unique to PostGIS, with no inheritance from classes with their pre-existing setup and tests, you can use it as a pattern:
- https://github.com/geotools/geotools/blob/6ff4135e3ddcd1890ff53aa5a676075afa4af54f/modules/plugin/jdbc/jdbc-postgis/src/test/java/org/geotools/data/postgis/PostGISHStoreTestSetup.java#L24
- https://github.com/geotools/geotools/blob/6ff4135e3ddcd1890ff53aa5a676075afa4af54f/modules/plugin/jdbc/jdbc-postgis/src/test/java/org/geotools/data/postgis/PostGISHStoreOnlineTest.java

**Review by aaime — CHANGES_REQUESTED** [review]

File: modules/plugin/geopkg/src/test/java/org/geotools/geopkg/GeoPkgDataStoreContentsOnlyTest.java (hunk 3 above)
Comment (ydeliorman) [reply modules/plugin/geopkg/src/test/java/org/geotools/geopkg/GeoPkgDataStoreContentsOnlyTest.java]: Fixed.

File: modules/plugin/geopkg/src/test/java/org/geotools/geopkg/GeoPkgDataStoreFactoryTestSetup.java (hunk 4 above)
Comment (ydeliorman) [reply modules/plugin/geopkg/src/test/java/org/geotools/geopkg/GeoPkgDataStoreFactoryTestSetup.java]: thanks for the pointer, fixed.

**Comment by aaime** [comment]
@ydeliorman merged, please backport as you see fit.

**Pull request merged**